Index: xapian-core/api/expanddecider.cc
===================================================================
--- xapian-core/api/expanddecider.cc	(revision 17087)
+++ xapian-core/api/expanddecider.cc	(working copy)
@@ -45,4 +45,16 @@
     return i == rejects.end();
 }
 
+bool    
+ExpandDeciderFilterPrefix::operator()(const string &term) const
+{
+    /* Filter is invalid when prefix parameter is a empty string, it means this
+     * filetr will do nothing. */
+    if (length <= 0) {
+        return true;
+    } 
+    string p = term.substr(0, length);
+    return p != prefix;    
 }
+
+}
Index: xapian-core/include/xapian/expanddecider.h
===================================================================
--- xapian-core/include/xapian/expanddecider.h	(revision 17087)
+++ xapian-core/include/xapian/expanddecider.h	(working copy)
@@ -96,6 +96,24 @@
     virtual bool operator()(const std::string &term) const;
 };
 
+class XAPIAN_VISIBILITY_DEFAULT ExpandDeciderFilterPrefix : public ExpandDecider {
+    std::string prefix;
+    unsigned length;
+
+  public:
+    /** The parameter specify the prefix of terms to be rejected
+     *  @param prefix_    Prefix of the term to reject. empty string means
+     *                    do nothing
+     */
+    ExpandDeciderFilterPrefix(const std::string &prefix_)
+	: prefix(prefix_) 
+    {   
+	length = prefix.length();
+    }   
+
+    virtual bool operator() (const std::string &term) const;
+};
+
 }
 
 #endif // XAPIAN_INCLUDED_EXPANDDECIDER_H
